kafka-instrumentation/src/main/java/dev/responsive/kafka/clients/EndOffsetsPoller.java

```diff
@@ -0,0 +1,220 @@
+  private final Metrics metrics;
+  private final Map<String, Object> configs;
+  private final Factories factories;
+  private ScheduledExecutorService executor;
+  private ScheduledFuture<?> pollFuture;
+  private Consumer<?, ?> consumer;
+
+  public EndOffsetsPoller(
+      final Map<String, ?> configs,
+      final Metrics metrics
+  ) {
+    this(configs, metrics, new Factories() {});
+  }
+
+  EndOffsetsPoller(
+      final Map<String, ?> configs,
+      final Metrics metrics,
+      final Factories factories
+  ) {
+    // TODO: make sure to use a read-committed consumer
+    this.configs = consumerConfigs(Objects.requireNonNull(configs));
+    this.metrics = Objects.requireNonNull(metrics);
+    this.factories = Objects.requireNonNull(factories);
+  }
+
+  private Map<String, Object> consumerConfigs(final Map<String, ?> providedConfigs) {
+    final HashMap<String, Object> configs = new HashMap<>(providedConfigs);
+    configs.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, BytesDeserializer.class);
+    configs.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, BytesDeserializer.class);
+    return Map.copyOf(configs);
+  }
+
+  public ResponsiveConsumer.Listener addForThread(final String threadId) {
+    LOGGER.info("add end offset metrics for thread {}", threadId);
+    final var tm = new ThreadMetrics(threadId);
+    synchronized (this) {
+      assert !threadIdToMetrics.containsKey(threadId);
+      if (threadIdToMetrics.isEmpty()) {
+        initPoller();
+        threadIdToMetrics.put(threadId, tm);
+      }
+    }
+    return tm;
+  }
+
+  public void removeForThread(final String threadId) {
+    final ThreadMetrics tm;
+    synchronized (this) {
+      tm = threadIdToMetrics.remove(threadId);
+      if (tm != null) {
+        tm.cleanup();
+        if (threadIdToMetrics.isEmpty()) {
+          stopPoller();
+        }
+      }
+    }
+  }
+
+  private void pollEndOffsets() {
+    final Set<TopicPartition> partitions = new HashSet<>();
+    final Collection<ThreadMetrics> threadMetrics;
+    synchronized (this) {
+      threadMetrics = List.copyOf(this.threadIdToMetrics.values());
+    }
+    for (final var tm : threadMetrics) {
+      partitions.addAll(tm.endOffsets.keySet());
+    }
+    final var endOffsets = consumer.endOffsets(partitions);
+    threadMetrics.forEach(tm -> tm.update(endOffsets));
+  }
+
+  private void initPoller() {
+    LOGGER.info("init end offsets poller");
+    assert consumer == null;
+    assert executor == null;
+    consumer = factories.createConsumer(configs);
+    executor = factories.createExecutor();
+    pollFuture = executor.scheduleAtFixedRate(this::pollEndOffsets, 0, 30, TimeUnit.SECONDS);
+  }
+
+  private void stopPoller() {
+    LOGGER.info("stopping end offsets poller");
+    try {
+      consumer.close(Duration.ofNanos(0));
+    } catch (final Exception e) {
+      LOGGER.warn("error closing consumer", e);
+    }
+    consumer = null;
```

Comment:
I believe there's a race here with `pollEndOffsets` - we could set this to null/close it and then after that a previously scheduled `pollEndOffsets` call gets triggered and line 120 would throw an NPE (or already closed exception)

is it really so bad to just synchronize everything in this class so we don't have to think about races? how often are we starting/stopping consumers?

Reply:
Ah good catch. I'd rather not block this on the client call - but I'll fix this race.